pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy

```diff
@@ -733,41 +730,122 @@ class ModelValidatorImpl implements ModelValidator {
     boolean validateElement(@Nonnull ModelASTStages stages) {
         boolean valid = true
 
-        if (stages.stages.isEmpty()) {
-            errorCollector.error(stages, Messages.ModelValidatorImpl_NoStages())
+        if (stages.stages == null) {
+            errorCollector.error(stages, Messages.ModelValidatorImpl_RequiredSection("stages"))
             valid = false
-        }
+        } else {
+            if (stages.stages.isEmpty()) {
+                errorCollector.error(stages, Messages.ModelValidatorImpl_NoStages())
+                valid = false
+            }
 
-        def stageNames = stages.stages.collect { s ->
-            s.name
-        }
+            def stageNames = stages.stages.collect { s ->
+                s.name
+            }
 
-        stageNames.findAll { stageNames.count(it) > 1 }.unique().each { sn ->
-            errorCollector.error(stages, Messages.ModelValidatorImpl_DuplicateStageName(sn))
-            valid = false
+            stageNames.findAll { stageNames.count(it) > 1 }.unique().each { sn ->
+                errorCollector.error(stages, Messages.ModelValidatorImpl_DuplicateStageName(sn))
+                valid = false
+            }
         }
 
         return validateFromContributors(stages, valid)
     }
 
     boolean validateElement(@Nonnull ModelASTParallel parallel) {
-        return true;
+        return validateElement((ModelASTStages)parallel)
```

Comment:
I always forget how casting and method signatures work - I assume this does what it's meant to since otherwise you'd end up with `StackOverflowError`s. =)

Reply:
Yes. :)